lib/src/cancelable_operation.dart

```diff
@@ -124,55 +121,51 @@ class CancelableOperation<T> {
   /// returned by [cancel], then completes to [cancellationValue].
   Future<T?> valueOrCancellation([T? cancellationValue]) {
     var completer = Completer<T?>.sync();
-    value.then((result) => completer.complete(result),
-        onError: completer.completeError);
+    value.then(completer.complete, onError: completer.completeError);
 
-    _completer._cancelMemo.future.then((_) {
+    _completer._cancelCompleter.future.then((_) {
       completer.complete(cancellationValue);
     }, onError: completer.completeError);
 
     return completer.future;
   }
 
-  /// Registers callbacks to be called when this operation completes.
+  /// Creates a new cancelable operation to be completed
+  /// when this operation completes or is cancelled.
   ///
-  /// [onValue] and [onError] behave in the same way as [Future.then].
+  /// The [onValue] and [onError] callbacks behave in the same way as
+  /// for [Future.then], and the result of those callbacks is used to complete
+  /// the returned cancelable operation.
   ///
-  /// If [onCancel] is provided, and this operation is canceled, the [onCancel]
-  /// callback is called and the returned operation completes with the result.
+  /// If [onCancel] is provided, and the this operation is canceled,
+  /// the [onCancel] callback is called and the returned operation completes
+  /// with the result returned by that call.
   ///
-  /// If [onCancel] is not given, and this operation is canceled, then the
-  /// returned operation is canceled.
+  /// If [onCancel] is not provided, and this operation is canceled, then the
+  /// returned operation is also canceled.
   ///
-  /// If [propagateCancel] is `true` and the returned operation is canceled then
-  /// this operation is canceled. The default is `false`.
+  /// If [propagateCancel] is `true` and the returned operation is canceled
+  /// then this operation is canceled. The default is `false`.
   CancelableOperation<R> then<R>(FutureOr<R> Function(T) onValue,
       {FutureOr<R> Function(Object, StackTrace)? onError,
       FutureOr<R> Function()? onCancel,
       bool propagateCancel = false}) {
     final completer =
         CancelableCompleter<R>(onCancel: propagateCancel ? cancel : null);
 
-    valueOrCancellation().then((T? result) {
-      if (!completer.isCanceled) {
-        if (isCompleted && !isCanceled) {
-          assert(result is T);
-          completer.complete(Future.sync(() => onValue(result as T)));
-        } else if (onCancel != null) {
-          completer.complete(Future.sync(onCancel));
-        } else {
-          completer._cancel();
-        }
-      }
-    }, onError: (Object error, StackTrace stackTrace) {
-      if (!completer.isCanceled) {
-        if (onError != null) {
-          completer.complete(Future.sync(() => onError(error, stackTrace)));
-        } else {
-          completer.completeError(error, stackTrace);
-        }
+    if (!isCanceled) {
+      value
+          .then(onValue, onError: onError)
```

Comment:
This is what is causing a few tests to fail. Before the `onValue` would not be called if the returned operation is canceled, but now it only matters whether this operation is canceled.

Reply:
Good point. I'll fix that.